packages/nextjs/src/server/clerkMiddleware.ts

```diff
@@ -178,6 +191,45 @@ export const clerkMiddleware: ClerkMiddleware = (...args: unknown[]): any => {
       return handlerResult;
     });
 
+    const nextMiddleware: NextMiddleware = async (request, event) => {
+      if (!ephemeralMode) {
+        return await baseNextMiddleware(request, event);
+      }
+
+      const ephemeralParams = unpackEphemeralQueryParams(request);
+
+      if (ephemeralParams) {
+        ephemeralAccount = ephemeralParams;
+
+        const response = new NextResponse(null, {
+          status: 307,
+          headers: { location: `${request.nextUrl.protocol}//${request.nextUrl.host}` },
+        });
+
+        const options = {
+          expires: Number(ephemeralAccount.expiresAt) * 1000,
+        };
+
+        response.cookies.set(constants.Cookies.EphemeralExpiresAt, ephemeralAccount.expiresAt.toString(), options);
+        response.cookies.set(constants.Cookies.EphemeralPublishableKey, ephemeralAccount.publishableKey, options);
+        response.cookies.set(constants.Cookies.EphemeralSecretKey, ephemeralAccount.secretKey, options);
+
+        return response;
+      }
```

Comment:
Can we get rid of the query param -> redirect -> cookies -> redirect flow? We do have a way to pass keys from the middleware to the app server as of a few months back, so potentially we can have a flow that originates in ClerkProvider, does a single redirect, and the middleware sets up the options so they're available to the subsequent request.

Reply:
Yeah, I think we can probably get this down to a single redirect - will do some testing to make sure that works as intended.